### ISSUE TITLE
NeatCanvasFeatures tubular look (gradient) for non-coding transcripts

### ISSUE THREAD

**Issue by mara-sangiovanni** [GMOD/jbrowse#1311]
Dear JBrowse developers and maintainers,
I have a question for you, I hope this is the right place to ask.

I'm building a JBrowse instance to visualize the REFSEQ genome/annotation of a marine species, Ciona intestinalis.
I'm using the latest official REFSEQ assembly + annotation version (GCF_000224145.3).

I'm using JBrowse 1.16.1 to exploit the possibility of coloring non-coding transcripts in a different way than coding ones, when using NeatCanvasFeatures. 
I'm also using NeatCanvasFeatures to have the tubular view, obtained with the "gradient":true option.

However there is no tubular gradient visualization for non-coding transcripts, and this is not really nice to look at, especially when a same gene has both coding and non-coding features  (see attached images).

![jbrowse_1](https://user-images.githubusercontent.com/47454675/52556066-bb08ee80-2deb-11e9-8058-3e73e0370ffa.png)

As far as I understand, this is due to the different gff features, since whenever the structure is the following everything is right:
gene -> mRNA -> exon -> CDS

However there are a lot of transcribed (but not translated into protein) features (you can give a look to the gff file I attached) that do not follow the above structure, since they are missing the CDS part, like in:

gene -> transcript -> exon
gene -> lnc_RNA -> exon
pseudogene -> transcript -> exon

Now, for the non coding features that DO NOT start with "gene" (pseudogene, cDNaMatch,...), the color is not really managed and is the same of the gene feature (i.e. style -> color). For all the others, the color is given by the style -> unprocessedTranscriptColor (you can find the settings in the attached trackList.json).

This will result in three different visualizations: one for the coding genes (tubular with the style color) and two for the non-coding (no tubular looks and two different colors).
So LOC113474495 (lcn_RNA) is green, LOC100185935 (pseudogene ) is steelblue, LOC100181953 (coding gene) is tubular + steelblue.

![jbrowse_2](https://user-images.githubusercontent.com/47454675/52556075-bcd2b200-2deb-11e9-96b7-eeece398483b.png)

The last but not the least, for the non coding transcripts that DO NOT start with gene-> there is no exon structure visualization, thus resulting in another difference among these two types of non-coding features (LOC100183649, pseudogene with exons, see image)
![jbrowse_3](https://user-images.githubusercontent.com/47454675/52558816-dd9f0580-2df3-11e9-8c6c-7d33dacf2e19.png)

Concluding, it would be nice to have a homogeneous view of coding and/or non-coding features. 
In particular:  
1)  can I get the tubular look for all the features? I remember that in previous versions of JBrowse this was possible. Am I doing something wrong? Can you point me in the right direction?

2) is there a way to assign the same color to ALL non-coding features  (it would be nice to have a style for all of them) so that (apart from the tubular look) all that is coding has a color, and all that is non-coding has another one?
 
3) following the same idea of item 2) , is it possible to show the exon structure also for non coding features that do not belong to the gene->... schema?

Thanks in advance, sorry for the verbosity
Mara

[TrackList.json.txt](https://github.com/GMOD/jbrowse/files/2850544/TrackList.json.txt)
[test_Ciona_REFSEQ_all.gff3.txt](https://github.com/GMOD/jbrowse/files/2850721/test_Ciona_REFSEQ_all.gff3.txt)

**Comment by cmdcolin** [comment]
1. The code for neat features appears to have always checked for "type"==="CDS" for coloring gradients
2. I think your config is essentially at this juncture, you have steelblue for processed transcript and green for noncoding/unprocessed transcript
3. this is not accepted by the default glyph feature. this was started to be discussed here but only ended with allowing noncoding transcripts to have gene parents https://github.com/GMOD/jbrowse/issues/1106

**Comment by mara-sangiovanni** [comment]
Dear Colin,
thanks for your answer. I try to explain better what I was asking:

1) This is the visualization of the same regions of the genome in JBrowse 1.14.2. Although exon visualization is somehow weird,  tubular look was available for all types of features. That's why I was asking how to get the same in the current version.

![jbrowse_4](https://user-images.githubusercontent.com/47454675/52632337-8b2c1a80-2ec1-11e9-9a4a-b82f50269d1e.png)

![jbrowse_5](https://user-images.githubusercontent.com/47454675/52632333-88c9c080-2ec1-11e9-8a0f-6cb5b139928d.png)

2)The problem is that I have long noncoding (LOC113474495) in green and the pseudogene (LOC100185935) in steelblue (determined by the coding genes color):  so two non coding, in two different colors. 
I could assign the same color to ALL features (coding and non coding), but then part of the non-coding will not be tubular (hence question n. 1).

So as far as I understood from your answers the only possibility to have a consistent view of the data is to have everything flat and with the same color: this is a little bit sad, since the NeatCanvasFeatures tubular look is way better than the flat. 

I hope that I better explained the situation,

Thanks again,
Mara

**Comment by mara-sangiovanni** [comment]
So no other suggestions? :(

BTW, if you're wondering why I don't use anymore JBrowse 1.14.2, is because it's not displaying the intron-exon structure of noncoding transcripts:

JBrowse 1.14.2 -> tubular look ma no structure

![jbrowse_8](https://user-images.githubusercontent.com/47454675/52863082-2d086d00-3137-11e9-8a6f-9c81e014f32a.png)

JBrowse 1.16.1 -> intron exon structure ma no tubular look

![jbrowse_6](https://user-images.githubusercontent.com/47454675/52863083-2d086d00-3137-11e9-8eb7-c8f19bef6f1c.png)

**Comment by cmdcolin** [comment]
I pushed some changes to the "dev" branch which I think will help

If you want to check it out, you can clone the github repo and run setup.sh and it should be able to be tested

**Comment by cmdcolin** [comment]
Oh and also checkout the dev branch instead of master too when testing this. Master is updated when we do a new release

**Comment by mara-sangiovanni** [comment]
Dear Colin,
thanks a lot! Things are working perfectly with the changes you made.
I used the dev branch with the same input and conf files I used before, and now the result is the one I was expecting.
I attach you the screenshots of how things look like now:

![jbrowse_11](https://user-images.githubusercontent.com/47454675/52958904-5d4d4700-3395-11e9-9276-c5bd036ccf0b.png)

![jbrowse_10](https://user-images.githubusercontent.com/47454675/52958903-5b838380-3395-11e9-9dd6-73bbbf834ee0.png)

Thanks again for the support,
Mara
